File: daemon-mac/Sources/FocusLockDaemon/FamilyEnforcementService.swift
import Foundation

/// Holds the currently-effective cloud rules and exposes a union view that
/// the enforcement loop applies alongside any active focus session.
///
/// Phase 2.3 semantics:
///   • block_now    → blocked domains + processes added to the union.
///   • unblock_all  → kill-switch; family-side enforcement suppressed while present.
///   • schedule     → stored and surfaced in status but cron evaluation is
///                    deferred to Phase 2.4.
///
/// Rules are cached to family-rules.json so daemon restart re-applies the
/// last-known state before the WebSocket reconnects.
final class FamilyEnforcementService {
    private static let cachePath = "/Library/Application Support/FocusLock/family-rules.json"

    private let lock = NSLock()
    private var rules: [String: CloudRule] = [:]
    private var listeners: [() -> Void] = []
    private let signer: IntegritySigner
    private let audit: ParentAuditService
    // Compiled cron evaluators keyed by rule ID. Rebuilt as rules change so
    // we don't re-parse on every tick.
    private var cron: [String: CronEvaluator] = [:]

    private let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.outputFormatting = [.prettyPrinted, .sortedKeys]
        return e
    }()
    private let decoder = JSONDecoder()

    init(signer: IntegritySigner, audit: ParentAuditService) {
        self.signer = signer
        self.audit = audit
        loadCache()
    }

    func onRulesChanged(_ cb: @escaping () -> Void) {
        lock.lock(); listeners.append(cb); lock.unlock()
    }

    private func fire() {
        lock.lock()
        let copy = listeners
        lock.unlock()
        for l in copy { l() }
    }

    var hasActiveBlocks: Bool {
        lock.lock(); defer { lock.unlock() }
        if hasUnblockAllLocked() { return false }
        let now = Date()
        // Mirror union(): collect block_now/schedule targets, then lift
        // anything covered by a live unblock_specific rule. If nothing
        // survives, no blocks are actually enforced.
        var domains = Set<String>()
        var procs   = Set<String>()
        for r in rules.values where r.active {
            if r.targetDomains.isEmpty && r.targetApps.isEmpty { continue }
            let active = r.kind == "block_now"
                || (r.kind == "schedule" && (cron[r.id]?.matches(now) ?? false))
            guard active else { continue }
            for d in r.targetDomains where !d.isEmpty { domains.insert(d) }
            for a in r.targetApps    where !a.isEmpty { procs.insert(a) }
        }
        if domains.isEmpty && procs.isEmpty { return false }
        for r in rules.values where r.active && r.kind == "unblock_specific" {
            if let exp = r.expiresAt, !isFuture(exp) { continue }
            for d in r.targetDomains { domains.remove(d) }
            for a in r.targetApps    { procs.remove(a) }
            if domains.isEmpty && procs.isEmpty { return false }
        }
        return !domains.isEmpty || !procs.isEmpty
    }

    func union() -> (domains: [String], processes: [String]) {
        lock.lock(); defer { lock.unlock() }
        if hasUnblockAllLocked() { return ([], []) }
        let now = Date()
        var domains = Set<String>()
        var procs   = Set<String>()
        for r in rules.values where r.active {
            let active = r.kind == "block_now"
                || (r.kind == "schedule" && (cron[r.id]?.matches(now) ?? false))
            guard active else { continue }
            for d in r.targetDomains where !d.isEmpty { domains.insert(d) }
            for a in r.targetApps    where !a.isEmpty { procs.insert(a) }
        }
        // Phase 3.2: unblock_specific lifts the matching items as long as the
        // rule hasn't expired. We treat a nil expiresAt defensively (the server
        // filters out expired rules on the wire, so we shouldn't see one here,
        // but if we do we treat it as live).
        for r in rules.values where r.active && r.kind == "unblock_specific" {
            if let exp = r.expiresAt, !isFuture(exp) { continue }
            for d in r.targetDomains { domains.remove(d) }
            for a in r.targetApps    { procs.remove(a) }
        }
        return (Array(domains), Array(procs))
    }

    func snapshot() -> [FamilyRuleSummary] {
        lock.lock(); defer { lock.unlock() }
        return rules.values.map { r in
            FamilyRuleSummary(
                id: r.id, kind: r.kind,
                targetApps: r.targetApps, targetDomains: r.targetDomains,
                scheduleCron: r.scheduleCron, createdAt: r.createdAt,
                expiresAt: r.expiresAt)
        }
    }

    // ── Mutators ───────────────────────────────────────────────────────────

    func upsert(_ rule: CloudRule) {
        if rule.id.isEmpty { return }
        var changed = false
        lock.lock()
        let existing = rules[rule.id]
        if existing == nil || !equalLocked(existing!, rule) {
            rules[rule.id] = rule
            rebuildCronLocked()
            saveCacheLocked()
            changed = true
        }
        lock.unlock()
        if changed {
            fputs("[family] rule upsert id=\(rule.id) kind=\(rule.kind)\n", stderr)
            fire()
        }
    }

    func remove(ruleId: String) {
        var changed = false
        lock.lock()
        if rules.removeValue(forKey: ruleId) != nil {
            rebuildCronLocked()
            saveCacheLocked()
            changed = true
        }
        lock.unlock()
        if changed {
            fputs("[family] rule removed id=\(ruleId)\n", stderr)
            fire()
        }
    }

    func replace(_ newRules: [CloudRule]) {
        lock.lock()
        rules = Dictionary(uniqueKeysWithValues: newRules
            .filter { !$0.id.isEmpty }
            .map { ($0.id, $0) })
        rebuildCronLocked()
        saveCacheLocked()
        lock.unlock()
        fputs("[family] rules replaced count=\(newRules.count)\n", stderr)
        fire()
    }

    func clearAll() {
        var changed = false
        lock.lock()
        if !rules.isEmpty {
            rules.removeAll()
            cron.removeAll()
            signer.deleteSigned(path: Self.cachePath)
            changed = true
        }
        lock.unlock()
        if changed {
            fputs("[family] rules cleared\n", stderr)
            fire()
        }
    }

    // ── Internals ──────────────────────────────────────────────────────────

    /// Recomputes `cron` from `rules`. Caller holds the lock.
    private func rebuildCronLocked() {
        var next: [String: CronEvaluator] = [:]
        for r in rules.values {
            guard r.kind == "schedule", let cronStr = r.scheduleCron, !cronStr.isEmpty else { continue }
            if let c = CronEvaluator.parse(cronStr) {
                next[r.id] = c
            } else {
                fputs("[family] schedule rule \(r.id) has invalid cron '\(cronStr)' — ignored\n", stderr)
            }
        }
        cron = next
    }

    private func hasUnblockAllLocked() -> Bool {
        for r in rules.values where r.active && r.kind == "unblock_all" { return true }
        return false
    }

    private func equalLocked(_ a: CloudRule, _ b: CloudRule) -> Bool {
        return a.active == b.active
            && a.kind == b.kind
            && a.scheduleCron == b.scheduleCron
            && a.expiresAt == b.expiresAt
            && a.targetApps == b.targetApps
            && a.targetDomains == b.targetDomains
    }

    /// True iff the ISO-8601 timestamp parses and lies strictly in the future.
    /// Used to decide whether an unblock_specific rule is still in effect.
    private func isFuture(_ iso: String) -> Bool {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = f.date(from: iso) { return d > Date() }
        // Some emitters omit fractional seconds; retry without that option.
        f.formatOptions = [.withInternetDateTime]
        if let d = f.date(from: iso) { return d > Date() }
        return false
    }

    private func loadCache() {
        guard let data = signer.readVerified(path: Self.cachePath) else {
            if FileManager.default.fileExists(atPath: Self.cachePath) {
                fputs("[family] cache has no valid signature — discarding\n", stderr)
                audit.record(ParentAuditEvents.familyCacheTampered, detail: "family-rules.json")
            }
            return
        }
        do {
            let arr = try decoder.decode([CloudRule].self, from: data)
            rules = Dictionary(uniqueKeysWithValues: arr.filter { !$0.id.isEmpty }.map { ($0.id, $0) })
            rebuildCronLocked()
            fputs("[family] cache loaded count=\(rules.count)\n", stderr)
        } catch {
            fputs("[family] cache unreadable: \(error)\n", stderr)
        }
    }

    private func saveCacheLocked() {
        do {
            let arr = Array(rules.values)
            let data = try encoder.encode(arr)
            signer.writeSigned(path: Self.cachePath, data: data)
        } catch {
            fputs("[family] cache write failed: \(error)\n", stderr)
        }
    }
}

File: daemon-win/FocusLock.Daemon/Services/FamilyEnforcementService.cs
using System.Text.Json;
using FocusLock.Daemon.Models;
using Microsoft.Extensions.Logging;

namespace FocusLock.Daemon.Services;

/// <summary>
/// Holds the *currently effective* cloud rules for this device and exposes a
/// union view that the enforcement loop can apply alongside any active focus
/// session.
///
/// Rule semantics:
///   • <c>block_now</c>  → blocked domains + processes added to the union.
///   • <c>unblock_all</c> → kill-switch that empties the union and suppresses
///                          family-side enforcement until the rule is removed.
///   • <c>schedule</c>   → cron-evaluated every tick (see <see cref="CronEvaluator"/>);
///                          a matching window contributes the rule's domains and
///                          processes to the same union as <c>block_now</c>.
///
/// Persistence: rules are cached to <c>family-rules.json</c> so a daemon
/// restart re-applies the last-known state before the WebSocket reconnects.
/// </summary>
public sealed class FamilyEnforcementService
{
    private static readonly string StateDir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData),
        "FocusLock");

    private static readonly string CachePath = Path.Combine(StateDir, "family-rules.json");

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true,
        WriteIndented = true,
    };

    private readonly ILogger<FamilyEnforcementService> _log;
    private readonly IntegritySigner _signer;
    private readonly ParentAuditService _audit;
    private readonly object _lock = new();
    private Dictionary<string, CloudRule> _rules = new();
    // Compiled cron evaluators keyed by rule ID. Rebuilt as rules change so
    // we don't re-parse on every tick.
    private Dictionary<string, CronEvaluator> _cron = new();

    public event Action? RulesChanged;

    public FamilyEnforcementService(ILogger<FamilyEnforcementService> log, IntegritySigner signer, ParentAuditService audit)
    {
        _log = log;
        _signer = signer;
        _audit = audit;
        Directory.CreateDirectory(StateDir);
        LoadCache();
    }

    /// <summary>True when at least one rule is enforcing blocks on this device.</summary>
    public bool HasActiveBlocks
    {
        get
        {
            lock (_lock)
            {
                if (HasUnblockAll()) return false;
                var now = DateTime.Now;
                foreach (var r in _rules.Values)
                {
                    if (!r.Active) continue;
                    if (r.TargetDomains.Count == 0 && r.TargetApps.Count == 0) continue;
                    if (r.Kind == "block_now") return true;
                    if (r.Kind == "schedule" && _cron.TryGetValue(r.Id, out var c) && c.Matches(now))
                        return true;
                }
                return false;
            }
        }
    }

    public (IReadOnlyCollection<string> Domains, IReadOnlyCollection<string> Processes) GetUnion()
    {
        lock (_lock)
        {
            if (HasUnblockAll())
                return (Array.Empty<string>(), Array.Empty<string>());

            var now = DateTime.Now;
            var domains = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            var procs   = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            foreach (var r in _rules.Values)
            {
                if (!r.Active) continue;
                bool active = r.Kind == "block_now"
                    || (r.Kind == "schedule"
                        && _cron.TryGetValue(r.Id, out var c) && c.Matches(now));
                if (!active) continue;
                foreach (var d in r.TargetDomains) if (!string.IsNullOrWhiteSpace(d)) domains.Add(d);
                foreach (var a in r.TargetApps)    if (!string.IsNullOrWhiteSpace(a)) procs.Add(a);
            }
            return (domains, procs);
        }
    }

    public IReadOnlyList<FamilyRuleSummary> Snapshot()
    {
        lock (_lock)
        {
            return _rules.Values.Select(r => new FamilyRuleSummary
            {
                Id            = r.Id,
                Kind          = r.Kind,
                TargetApps    = new List<string>(r.TargetApps),
                TargetDomains = new List<string>(r.TargetDomains),
                ScheduleCron  = r.ScheduleCron,
                CreatedAt     = r.CreatedAt,
            }).ToList();
        }
    }

    // ── Mutators called by CloudSyncService ────────────────────────────────

    public void Upsert(CloudRule rule)
    {
        if (string.IsNullOrEmpty(rule.Id)) return;
        bool changed;
        lock (_lock)
        {
            changed = !_rules.TryGetValue(rule.Id, out var existing) || !RuleEquals(existing, rule);
            _rules[rule.Id] = rule;
            if (changed) { RebuildCronLocked(); SaveCache(); }
        }
        if (changed)
        {
            _log.LogInformation("Family rule upserted: id={Id} kind={Kind} domains={D} apps={A}",
                rule.Id, rule.Kind, rule.TargetDomains.Count, rule.TargetApps.Count);
            RulesChanged?.Invoke();
        }
    }

    public void Remove(string ruleId)
    {
        bool changed;
        lock (_lock)
        {
            changed = _rules.Remove(ruleId);
            if (changed) { RebuildCronLocked(); SaveCache(); }
        }
        if (changed)
        {
            _log.LogInformation("Family rule removed: id={Id}", ruleId);
            RulesChanged?.Invoke();
        }
    }

    public void Replace(IEnumerable<CloudRule> rules)
    {
        lock (_lock)
        {
            _rules = rules
                .Where(r => !string.IsNullOrEmpty(r.Id))
                .ToDictionary(r => r.Id, r => r);
            RebuildCronLocked();
            SaveCache();
        }
        _log.LogInformation("Family rules replaced: count={Count}", _rules.Count);
        RulesChanged?.Invoke();
    }

    public void Clear()
    {
        lock (_lock)
        {
            if (_rules.Count == 0) return;
            _rules.Clear();
            _cron.Clear();
            _signer.DeleteSigned(CachePath);
        }
        _log.LogInformation("Family rules cleared");
        RulesChanged?.Invoke();
    }

    // ── Internals ──────────────────────────────────────────────────────────

    /// <summary>Recomputes <see cref="_cron"/> from <see cref="_rules"/>. Caller holds the lock.</summary>
    private void RebuildCronLocked()
    {
        var next = new Dictionary<string, CronEvaluator>();
        foreach (var r in _rules.Values)
        {
            if (r.Kind != "schedule" || string.IsNullOrWhiteSpace(r.ScheduleCron)) continue;
            var c = CronEvaluator.Parse(r.ScheduleCron);
            if (c != null) next[r.Id] = c;
            else _log.LogWarning("Schedule rule {Id} has invalid cron '{Cron}' — ignored", r.Id, r.ScheduleCron);
        }
        _cron = next;
    }

    private bool HasUnblockAll()
    {
        foreach (var r in _rules.Values)
            if (r.Active && r.Kind == "unblock_all") return true;
        return false;
    }

    private static bool RuleEquals(CloudRule a, CloudRule b) =>
        a.Active == b.Active &&
        a.Kind == b.Kind &&
        a.ScheduleCron == b.ScheduleCron &&
        a.TargetApps.SequenceEqual(b.TargetApps) &&
        a.TargetDomains.SequenceEqual(b.TargetDomains);

    private void LoadCache()
    {
        var bytes = _signer.ReadVerified(CachePath);
        if (bytes == null)
        {
            if (File.Exists(CachePath))
            {
                _log.LogWarning("Family rule cache has no valid signature — discarding");
                _audit.Record(ParentAuditEvents.FamilyCacheTampered, detail: "family-rules.json");
            }
            return;
        }
        try
        {
            var rules = JsonSerializer.Deserialize<List<CloudRule>>(bytes, JsonOpts);
            if (rules == null) return;
            _rules = rules.Where(r => !string.IsNullOrEmpty(r.Id)).ToDictionary(r => r.Id, r => r);
            RebuildCronLocked();
            _log.LogInformation("Loaded {Count} cached family rules", _rules.Count);
        }
        catch (Exception ex)
        {
            _log.LogWarning(ex, "Family rule cache unreadable — ignoring");
        }
    }

    private void SaveCache()
    {
        try
        {
            var bytes = JsonSerializer.SerializeToUtf8Bytes(_rules.Values.ToList(), JsonOpts);
            _signer.WriteSigned(CachePath, bytes);
        }
        catch (Exception ex)
        {
            _log.LogWarning(ex, "Failed to persist family rule cache");
        }
    }
}
